pkg/controllers/gateway/gateway_controller_test.go

```diff
@@ -0,0 +1,977 @@
+				t.Errorf("reconcileGateway() clusters = %v, want %v", clusters, tt.wantClusters)
+			}
+			if requeue != tt.wantRequeue {
+				t.Errorf("reconcileGateway() requeue = %v, want %v", requeue, tt.wantRequeue)
+			}
+		})
+	}
+}
+
+func TestGatewayReconciler_reconcileTLS(t *testing.T) {
+	type fields struct {
+		Client client.Client
+		Scheme *runtime.Scheme
+	}
+	type args struct {
+		upstreamGateway *v1beta1.Gateway
+		gateway         *v1beta1.Gateway
+		managedHosts    []v1alpha1.ManagedHost
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		want    []v1.Object
+		wantErr bool
+	}{
+		{
+			name: "secret synced downstream",
+			fields: fields{
+				Client: test.GetValidTestClient(getValidCertificateSecret(test.ValidTestHostname)),
+				Scheme: test.GetValidTestScheme(),
+			},
+			args: args{
+				upstreamGateway: &v1beta1.Gateway{
+					ObjectMeta: v1.ObjectMeta{
+						Namespace: test.Namespace,
+					},
+				},
+				gateway: &v1beta1.Gateway{
+					Spec: v1beta1.GatewaySpec{
+						Listeners: []v1beta1.Listener{
+							{
+								Name:     test.ValidTestHostname,
+								Hostname: getTestHostname(test.ValidTestHostname),
+								Protocol: v1beta1.HTTPSProtocolType,
+							},
+						},
+					},
+				},
+				managedHosts: []v1alpha1.ManagedHost{
+					{
+						Host: test.ValidTestHostname,
+					},
+				},
+			},
+			want:    []v1.Object{&getValidCertificateSecret(test.ValidTestHostname).Items[0]},
+			wantErr: false,
+		},
+		{
+			name: "no cert for HTTP listener",
+			fields: fields{
+				Client: test.GetValidTestClient(),
+				Scheme: test.GetValidTestScheme(),
+			},
+			args: args{
+				upstreamGateway: &v1beta1.Gateway{},
+				gateway: &v1beta1.Gateway{
+					Spec: v1beta1.GatewaySpec{
+						Listeners: []v1beta1.Listener{
+							{
+								Name:     test.ValidTestHostname,
+								Hostname: getTestHostname(test.ValidTestHostname),
+								Protocol: v1beta1.HTTPProtocolType,
+							},
+						},
+					},
+				},
+				managedHosts: []v1alpha1.ManagedHost{
+					{
+						Host: test.ValidTestHostname,
+					},
+				},
+			},
+			want:    []v1.Object{},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := &GatewayReconciler{
+				Client:       tt.fields.Client,
+				Scheme:       tt.fields.Scheme,
+				Certificates: test.NewTestCertificateService(tt.fields.Client),
+				Host:         test.NewTestHostService(tt.fields.Client),
+				Placement:    test.NewTestGatewayPlacer(),
+			}
+			got, err := r.reconcileTLS(context.TODO(), tt.args.upstreamGateway, tt.args.gateway, tt.args.managedHosts)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("reconcileTLS() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !verifyTLSSecretTestResultsAsExpected(got, tt.want, tt.args.upstreamGateway) {
+				t.Errorf("reconcileTLS() \ngot: \n%v \nwant: \n%v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestGatewayReconciler_reconcileDNSEndpoints(t *testing.T) {
```

Comment:
You need to rebase main, this no longer exists in the gateway controller.

Reply:
And a few more tests cases not don't make much sense. Removed